src/core/ddsi/tests/plist_leasedur.c

```diff
@@ -0,0 +1,480 @@
+  nn_xmsg_free (m);
+}
+
+CU_Test (ddsi_plist_leasedur, ser_sedp, .init = setup, .fini = teardown)
+{
+  ddsi_guid_t guid;
+  memset (&guid, 0, sizeof (guid));
+  struct nn_xmsg *m = nn_xmsg_new (gv.xmsgpool, &guid, NULL, 64, NN_XMSG_KIND_DATA);
+  CU_ASSERT_FATAL (m != NULL);
+  struct nn_xmsg_marker marker;
+  (void) nn_xmsg_append (m, &marker, 0);
+
+  ddsi_plist_t plist;
+  ddsi_plist_init_empty (&plist);
+  plist.qos.present |= QP_LIVELINESS;
+  plist.qos.liveliness.kind = DDS_LIVELINESS_MANUAL_BY_PARTICIPANT;
+  plist.qos.liveliness.lease_duration = 2041944443;
+  ddsi_plist_addtomsg_bo (m, &plist, 0, QP_LIVELINESS, DDSRT_BOSEL_BE, DDSI_PLIST_CONTEXT_ENDPOINT);
+  nn_xmsg_addpar_sentinel_bo (m, DDSRT_BOSEL_BE);
+
+  const uint8_t expected[] = { LL(1, 2,0x0abcdefb), SENTINEL };
+  const unsigned char *cdr = nn_xmsg_submsg_from_marker (m, marker);
+  CU_ASSERT (memcmp (expected, cdr, sizeof (expected)) == 0);
+  
+  ddsi_plist_fini (&plist);
+  nn_xmsg_free (m);
+}
+
+#define UDPLOCATOR(a,b,c,d,port) \
+  SER32BE (NN_LOCATOR_KIND_UDPv4), \
+  SER32BE(port), \
+  SER32BE(0),SER32BE(0),SER32BE(0), \
+  (a),(b),(c),(d)
+
+#define TEST_GUIDPREFIX_BYTES 7,7,3,4, 5,6,7,8, 9,10,11,12
+
+static void setup_and_start (void)
+{
+  setup ();
+  ddsi_set_deafmute (&gv, true, true, DDS_INFINITY);
+  rtps_start (&gv);
+  // Register the main thread, then claim it as spawned by Cyclone because the
+  // internal processing has various asserts that it isn't an application thread
+  // doing the dirty work
+  struct thread_state * const thrst = ddsi_lookup_thread_state ();
+  assert (thrst->state == THREAD_STATE_LAZILY_CREATED);
+  thrst->state = THREAD_STATE_ALIVE;
+  ddsrt_atomic_stvoidp (&thrst->gv, &gv);
+}
+
+static void stop_and_teardown (void)
+{
+  // Shutdown currently relies on sending packets to shutdown receiver threads
+  // handling individual sockets (this sometime causes issues with firewalls, too)
+  ddsi_set_deafmute (&gv, false, false, DDS_INFINITY);
+  // On shutdown there is an expectation that the thread was discovered dynamically.
+  // We overrode it in the setup code, we undo it now.
+  struct thread_state * const thrst = ddsi_lookup_thread_state ();
+  thrst->state = THREAD_STATE_LAZILY_CREATED;
+  rtps_stop (&gv);
+  teardown ();
+}
+
+static void ddsi_plist_leasedur_new_proxypp_impl (bool include_lease_duration)
+{
+  struct thread_state * const thrst = ddsi_lookup_thread_state ();
+  const uint32_t port = gv.loc_meta_uc.port;
+
+  // not static nor const: we need to patch in the port number
+  unsigned char pkt_header[] = {
+    'R', 'T', 'P', 'S', RTPS_MAJOR, RTPS_MINOR,
+    // vendor id: major 1 is a given
+    1, NN_VENDORID_MINOR_ECLIPSE,
+    // GUID prefix: first two bytes ordinarily have vendor id, so 7,7 is
+    // guaranteed to not be used locally
+    TEST_GUIDPREFIX_BYTES,
+    // DATA (0x15); flags (4 = dataflag + big-endian); octets-to-next-header = 0
+    // means it continues until the end
+    0x15, 4, 0,0,
```

Comment:
You could use the `SMID_DATA` macro here (and other macros for similar cases below).